src/System.Text.Primitives/System/Text/Encoders/UnicodeScalar.cs

```diff
@@ -0,0 +1,246 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics;
+using System.Globalization;
+
+namespace System.Buffers.Text
+{
+    /// <summary>
+    /// Represents a 24-bit Unicode scalar value.
+    /// A scalar value is any value in the range [U+0000..U+D7FF] or [U+E000..U+10FFFF].
+    /// </summary>
+    public struct UnicodeScalar : IComparable<UnicodeScalar>, IEquatable<UnicodeScalar>
+    {
+        /// <summary>
+        /// The Unicode Replacement Character U+FFFD.
+        /// </summary>
+        public static readonly UnicodeScalar ReplacementChar = new UnicodeScalar(0xFFFD);
+
+        /// <summary>
+        /// The integer value of this scalar.
+        /// </summary>
+        public readonly int Value; // = U+0000 if using default init
+
+        /// <summary>
+        /// Constructs a Unicode scalar from the given UTF-16 code point.
+        /// The code point must not be a surrogate.
+        /// </summary>
+        /// <param name="char"></param>
+        public UnicodeScalar(char @char)
+            : this((uint)@char)
+        {
+            // None of the APIs on this type are guaranteed to produce correct results
+            // if we don't validate the input during construction.
+
+            if (Utf8Util.IsLowWordSurrogate((uint)Value))
+            {
+                throw new ArgumentOutOfRangeException(
+                   message: "Value must be between U+0000 and U+D7FF, inclusive; or value must be between U+E000 and U+FFFF, inclusive.",
+                   paramName: nameof(@char));
+            }
+        }
+
+        /// <summary>
+        /// Constructs a Unicode scalar from the given value.
+        /// The value must represent a valid scalar.
+        /// </summary>
+        public UnicodeScalar(int value)
+            : this((uint)value)
+        {
+            // None of the APIs on this type are guaranteed to produce correct results
+            // if we don't validate the input during construction.
+
+            if (!IsValidScalar((uint)Value))
+            {
+                throw new ArgumentOutOfRangeException(
+                    message: "Value must be between U+0000 and U+D7FF, inclusive; or value must be between U+E000 and U+10FFFF, inclusive.",
+                    paramName: nameof(value));
+            }
+        }
+
+        // non-validating ctor for internal use
+        private UnicodeScalar(uint value)
+        {
+            Value = (int)value;
+        }
+
+        /// <summary>
+        /// Converts a <see cref="char"/> to a <see cref="UnicodeScalar"/>.
+        /// Identical to calling <see cref="UnicodeScalar.UnicodeScalar(char)"/>.
+        /// </summary>
+        public static explicit operator UnicodeScalar(char @char) => new UnicodeScalar(@char);
```

Comment:
Why do we need these explicit operators?

Reply:
I thought they'd be more convenient and result in nicer looking code than calling the ctor, but TBH I don't have a use case for it right now. Maybe I should just delete?